tests/ScopeTest.php

```diff
@@ -162,31 +193,64 @@ public function testContitionOnReferencedRecords()
             $this->assertEquals('LV', $u->get('country_code'));
         }
 
+        $user = clone $this->user;
+        
+        // users that have no ticket
+        $user->addCondition('Tickets/#', 0);
+        
+        $this->assertEquals(3, $user->action('count')->getOne());
+        
+        foreach ($user as $u) {
+            $this->assertTrue(in_array($u->get('name'), ['Alain', 'Aerton', 'Rubens']));
+        }
+        
         $country = clone $this->country;
 
-        // countries with no users
-        $country->addCondition('Users/!');
+        // countries with more than one user
+        $country->addCondition('Users/#', '>', 1);
 
         foreach ($country as $c) {
-            $this->assertEmpty($c->get('user_names'));
+            $this->assertEquals('BR', $c->get('code'));
         }
-
+        
         $country = clone $this->country;
 
-        // countries with any user
-        $country->addCondition('Users/?');
+        // countries with users that have ticket number 001
+        $country->addCondition('Users/Tickets/number', '001');
 
         foreach ($country as $c) {
-            $this->assertNotEmpty($c->get('user_names'));
+            $this->assertEquals('CA', $c->get('code'));
         }
+        
+        $country = clone $this->country;
 
+        // countries with users that have more than one ticket
+        $country->addCondition('Users/Tickets/#', '>', 1);
+        
+        foreach ($country as $c) {
+            $this->assertEquals('LV', $c->get('code'));
+        }
+        
         $country = clone $this->country;
 
-        // countries with more than one user
-        $country->addCondition('Users/#', '>', 1);
+        // countries with users that have any tickets
+        $country->addCondition('Users/Tickets/#');
+        
+        $this->assertEquals(2, $country->action('count')->getOne());
+        
+        foreach ($country as $c) {
+            $this->assertTrue(in_array($c->get('code'), ['LV', 'CA']));
+        }
+        
+        $country = clone $this->country;
+        
+        // countries with users that have no tickets
+        $country->addCondition('Users/Tickets/#', 0);
 
+        $this->assertEquals(2, $country->action('count')->getOne());
+        
         foreach ($country as $c) {
-            $this->assertEquals('BR', $c->get('code'));
+            $this->assertTrue(in_array($c->get('code'), ['FR', 'BR']));
         }
```

Comment:
I am still missing one test which will reference one Model twice

can we add field `restrictedToCountry` field to ticket model?

Reply:
Give me an example, please

Reply:
```
// users with tickets that have more than two users per country)
// test if User model can be referenced multiple times
$user->addCondition('Tickets/user/Country/Users/#', '>=', 2);
// add twice to test if generated query has no duplicate column names because of counting/# field
$user->addCondition('Tickets/user/Country/Users/#', '>', 2);
```

Reply:
I just included the test and it works.
No JOINs are used. It is simply a chain for exists -> exists -> exists with adding the relation of the tables.

Users with tickets whose user has country with more than 1 users.